### PR TITLE
Analyse Merge: charging_station in France

### PR DESCRIPTION
**Pull request by nlehuby**
Une analyse merge pour les stations de recharge de véhicules électriques, en prévision d'un [projet du mois](https://wiki.openstreetmap.org/wiki/FR:Project_of_the_month/bornes_vehicules_electriques).

Les données open data sont retraitées au préalable sur un autre repo, car le fichier contient des points de recharge alors que nous voulons des stations (qui regroupent potentiellement plusieurs points de recharge)

définitions des objets

Documentation du mapping sur le wiki : https://wiki.openstreetmap.org/wiki/France/data.gouv.fr/Bornes_de_Recharge_pour_V%C3%A9hicules_%C3%89lectriques

### CODE REVIEW

File: analysers/analyser_merge_charging_station_FR.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+##                                                                                                                                    ##
+## This program is distributed in the hope that it will be useful,                  ##
+## but WITHOUT ANY WARRANTY; without even the implied warranty of ##
+## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See     ##
+## the GNU General Public License for more details.                                        ##
+##                                                                                                                                    ##
+## You should have received a copy of the GNU General Public License     ##
+## along with this program.  If not, see <http://www.gnu.org/licenses/>. ##
+##                                                                                                                                    ##
+###########################################################################
```
Comment (frodrigo) [inline analysers/analyser_merge_charging_station_FR.py]: Please fix the code layout.

File: analysers/analyser_merge_charging_station_FR.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+class Analyser_Merge_Charging_station_FR(Analyser_Merge):
+    def __init__(self, config, logger=None):
+        Analyser_Merge.__init__(self, config, logger)
+        self.missing_official = self.def_class(item=9990, id=1, level=3, tags=['merge'],
+                                               title=T_('Car charging station not integrated'))
+        self.possible_merge = self.def_class(item=9991, id=3, level=3, tags=['merge'],
+                                             title=T_('Car charging station integration suggestion'))
+        self.update_official = self.def_class(item=9992, id=4, level=3, tags=['merge'],
+                                              title=T_('Car charging station update'))
+
```
Comment (frodrigo) [inline analysers/analyser_merge_charging_station_FR.py]: Please use the new code layout and user documentation.

See for example: https://github.com/osm-fr/osmose-backend/blob/master/analysers/analyser_merge_heritage_FR_merimee.py

File: analysers/analyser_merge_charging_station_FR.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        self.init(
+            u"https://transport.data.gouv.fr/datasets/fichier-consolide-des-bornes-de-recharge-pour-vehicules-electriques/",
```
Comment (frodrigo) [inline analysers/analyser_merge_charging_station_FR.py]: u"", no more required, we are now in python3 \o/

File: analysers/analyser_merge_charging_station_FR.py (hunk 2 above)
Comment (frodrigo) [reply analysers/analyser_merge_charging_station_FR.py]: See the result of the documentation items on http://beta.osmose.openstreetmap.fr

File: analysers/analyser_merge_charging_station_FR.py (hunk 1 above)
Comment (frodrigo) [reply analysers/analyser_merge_charging_station_FR.py]: Better, but still broken on Copyrights.

Comment (frodrigo) [reply analysers/analyser_merge_charging_station_FR.py]: Fixed

File: analysers/analyser_merge_charging_station_FR.py (hunk 2 above)
Comment (frodrigo) [reply analysers/analyser_merge_charging_station_FR.py]: I fixed the code layout.

File: analysers/analyser_merge_charging_station_FR.py (hunk 1 above)
Comment (nlehuby) [reply analysers/analyser_merge_charging_station_FR.py]: thanks

**Comment by frodrigo** [comment]
It's now ready for you?
Please add it also to configuration.

**Comment by nlehuby** [comment]
yes, it's ok for me.

File: osmose_config.py [hunk 4]
```diff
@@ -256,6 +256,7 @@ class gen(default_country):
     'osmosis_highway_motorway',
     'osmosis_highway_zone',
     'merge_shop_FR',
+    'merge_charging_station_FR',
```
Comment (frodrigo) [inline osmose_config.py]: Please, set it on local_db.

**Comment by frodrigo** [comment]
Thank you merged. But still not deployed.

**Comment by frodrigo** [comment]
The item number are not good. Should be 8410.

**Comment by nlehuby** [comment]
8410, 8411 and 8412 ?

**Comment by frodrigo** [comment]
Yes

**Comment by frodrigo** [comment]
It is not working
```
2020-01-29 03:26:06   error: error on analyse merge_charging_station_FR...
2020-01-29 03:26:06     Traceback (most recent call last):
2020-01-29 03:26:06       File "./osmose_run.py", line 230, in execc
2020-01-29 03:26:06         analyser_obj.analyser()
2020-01-29 03:26:06       File "/data/project/osmose/backend/analysers/Analyser_Osmosis.py", line 192, in analyser
2020-01-29 03:26:06         self.analyser_osmosis_common()
2020-01-29 03:26:06       File "/data/project/osmose/backend/analysers/Analyser_Merge.py", line 889, in analyser_osmosis_common
2020-01-29 03:26:06         table = self.load.run(self, self.parser, self.mapping, self.config.db_user, self.__class__.__name__.lower()[15:], self.analyser_version())
2020-01-29 03:26:06       File "/data/project/osmose/backend/analysers/Analyser_Merge.py", line 724, in run
2020-01-29 03:26:06         osmosis.run0((sql01_ref if mapping.osmRef != "NULL" else sql01_geo) % {"table":table, "x":self.x, "y":self.y, "where":self.formatCSVSelect(), "distinct": distinct, "order_by": order_by}, insertOfficial)
2020-01-29 03:26:06       File "/data/project/osmose/backend/analysers/Analyser_Osmosis.py", line 436, in run0
2020-01-29 03:26:06         self.run00(sql, callback)
2020-01-29 03:26:06       File "/data/project/osmose/backend/analysers/Analyser_Osmosis.py", line 427, in run00
2020-01-29 03:26:06         ret = callback(res)
2020-01-29 03:26:06       File "/data/project/osmose/backend/analysers/Analyser_Merge.py", line 701, in insertOfficial
2020-01-29 03:26:06         tags = mapping.generate.tagFactory(res)
2020-01-29 03:26:06       File "/data/project/osmose/backend/analysers/Analyser_Merge.py", line 805, in tagFactory
2020-01-29 03:26:06         tags_secondary = self.tagFactoryGroup(res, self.static2, self.mapping2)
2020-01-29 03:26:06       File "/data/project/osmose/backend/analysers/Analyser_Merge.py", line 795, in tagFactoryGroup
2020-01-29 03:26:06         r = colomn(res)
2020-01-29 03:26:06       File "/data/project/osmose/backend/analysers/analyser_merge_charging_station_FR.py", line 81, in <lambda>
2020-01-29 03:26:06         "opening_hours": lambda fields: guess_opening_hours(fields["accessibilité_grouped"]),
2020-01-29 03:26:06       File "/data/project/osmose/backend/analysers/analyser_merge_charging_station_FR.py", line 95, in guess_opening_hours
2020-01-29 03:26:06         if "24/24" in text_opening_hours:
2020-01-29 03:26:06     TypeError: argument of type 'NoneType' is not iterable
```

**Comment by nlehuby** [comment]
I can't reproduce :thinking: can this be a cache issue ?

**Comment by nlehuby** [comment]
I've just pushed a potential fix on #748

**Comment by frodrigo** [comment]
@jocelynj 2020-02-05 09:37:27       error: UPDATE ERROR france_local_db/Merge_Charging_station_FR : AUTH FAIL

**Comment by frodrigo** [comment]
Please can you check the results: http://osmose.openstreetmap.fr/fr/errors/?item=8410&useDevItem=all

**Comment by bmillemathias** [comment]
I see there is a lot of uppercase strings, the city for instance (coming from `ad_station`) or network (ex: IRVE SDEY), operator or owner.
Should they be downcased ?

**Comment by frodrigo** [comment]
Maybe lower case for operator.

I also see encoding error like `R‚seau Alterbase - Sauz‚ Vaussais - P“le M‚dical`.

**Comment by bmillemathias** [comment]
> I also see encoding error like `R‚seau Alterbase - Sauz‚ Vaussais - P“le M‚dical`.

this is coming from the csv, not the processing.

**Comment by nlehuby** [comment]
I'm not sure about the downcasing: we have a lot of abbreviations, I'm pretty sure we should keep as is `SDE07`, `SDED` (and all the `SDE something`: SDE stands for `syndicat d'énergie`), `SPIE`, `InEnergies`, etc

**Comment by frodrigo** [comment]
Change have to de done ? Are we ready ? We keep it hidden until project start ?

**Comment by nlehuby** [comment]
I did not have a chance to give it a proper review yet. I try to do it by the week-end.

**Comment by nlehuby** [comment]
I cannot see the doc panel on osmose website for these items, is it something wrong in the conf ?

**Comment by frodrigo** [comment]
It is an hidden item. It probably also hide the doc.

**Comment by nlehuby** [comment]
I have added more traps in the doc in #766 
It would be cool if we can have it deployed with French translations in March.

**Comment by frodrigo** [comment]
You can translate here https://www.transifex.com/openstreetmap-france/osmose/backend/

**Comment by nlehuby** [comment]
translation is ok

**Comment by nlehuby** [comment]
Can we make it public by the week-end ?

**Comment by jocelynj** [comment]
It is now enabled on osmose frontend.

**Comment by nlehuby** [comment]
Thanks

**Pull request merged**